### PR TITLE
[STORM-1157] Adding dynamic profiling for worker, restarting worker, jstack, heap dump, and profiling

### PR DESCRIPTION
**Pull request by kishorvpatil**
- Request restarting worker.
- Request jstack, 
- Request heap dump, 
- Start Profiling
   - Dump profiling
   - Stop Profiling

### CODE REVIEW

**Comment by revans2** [comment]
It looks good to me.  I am a little bit concerned that there is no way to turn this off, partially or all the way.  So for windows where there is currently no support for this we will end up trying to run a shell script that will fail, and if someone does not have/want commercial java support from Oracle, i.e. using open JDK, or IBM JDK, etc.  They have to make code modifications, not just config modifications to not show users buttons that will not work.

Could we put in something that could let users turn off part or all of this feature?  At a minimum in the UI, but it would be good if nimbus could also let users know that this particular feature is not currently configured on, and if the worker could see the config and know to add in ```-XX:+UnlockCommercialFeatures -XX:+FlightRecorder``` on the command line.

Other then that I am +1.

**Comment by kishorvpatil** [comment]
@revans2 Thank you for the feedback. Following changes should help make this feature completely optional and pluggable.   
   - Added `worker.profiler.enabled` to shutdown use of profilers - 
     - If disabled, UI simply removes the section **Profiling and Debugging** section. 
     - If disabled, the supervisor would not add "worker.profiler.childopts" to worker JVM options - hence avoiding issues with jvm that does not provide commercial JProfiler. 
   - Added `worker.profiler.command` to specify which script to use at supervisor level so that it can be customized for different profiler scheme.

**Comment by kishorvpatil** [comment]
Currently, this feature lacks support for windows and command-line interface.

**Comment by kishorvpatil** [comment]
dynamic_profiling_debugging_1
dynamic_profiling_debugging_2
dynamic_profiling_debugging_3

**Comment by revans2** [comment]
+1 it looks like a great addition.

File: docs/DYNAMIC_WORKER_PROFILING.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
+Dynamic Worker Profiling
```
Comment (zhuoliu) [inline docs/DYNAMIC_WORKER_PROFILING.md]: Should we move this doc file to docs/documentation directory?
Also the DYNAMIC_LOG_LEVEL_SETTINGS.md file.

**Comment by kishorvpatil** [comment]
@zhuoliu I moved docs under documentation location. Also, added pointers from main page.

File: storm-core/src/clj/backtype/storm/ui/core.clj [hunk 2]
```diff
@@ -986,6 +1021,127 @@
         (log-message "Setting topology " id " log config " new-log-config)
         (.setLogConfig nimbus id new-log-config)
         (json-response (log-config id) (m "callback")))))
+
+  (GET "/api/v1/topology/:id/profiling/start/:host-port/:timeout"
```
Comment (zhuoliu) [inline storm-core/src/clj/backtype/storm/ui/core.clj]: We may need to update the REST API doc for this.

File: storm-core/src/clj/backtype/storm/ui/core.clj [hunk 3]
```diff
@@ -986,6 +1021,127 @@
...
+           (json-response {"status" "ok"
+                           "id" host-port
+                           "timeout" timeout
+                           "dumplink" (worker-dump-link
+                                       host
+                                       port
+                                       id)}
+                          (m "callback")))))
+
+  (GET "/api/v1/topology/:id/profiling/stop/:host-port"
```
Comment (zhuoliu) [inline storm-core/src/clj/backtype/storm/ui/core.clj]: Same here and below.

**Comment by zhuoliu** [comment]
Just some nits about docs. Otherwise looks really good to me.

**Comment by kishorvpatil** [comment]
@zhuoliu I have updated the REST API documentation for all new endpoints.

**Comment by zhuoliu** [comment]
+1

File: STORM-UI-REST-API.md [hunk 4]
```diff
@@ -636,6 +636,201 @@ Sample response:
...
 
+## Profiling and Debugging GET Operations
+
+###  /api/v1/topology/:id/profiling/start/:host-port/:timeout (GET)
+
+Request to start profiler on worker with timeout. Returns status and link to profiler artifacts for worker.
+
+|Parameter |Value   |Description  |
+|----------|--------|-------------|
+|id   	   |String (required)| Topology Id  |
```
Comment (revans2) [inline STORM-UI-REST-API.md]: Do we want an `:id` here to be consistent with the other Parameters?

File: STORM-UI-REST-API.md [hunk 5]
```diff
@@ -636,6 +636,201 @@ Sample response:
...
+}
+```
+
+###  /api/v1/topology/:id/profiling/dumpprofile/:host-port (GET)
+
+Request to dump profiler recording on worker. Returns status and worker id for the request.
+
+|Parameter |Value   |Description  |
+|----------|--------|-------------|
+|id   	   |String (required)| Topology Id  |
```
Comment (revans2) [inline STORM-UI-REST-API.md]: Here too.

File: STORM-UI-REST-API.md [hunk 6]
```diff
@@ -636,6 +636,201 @@ Sample response:
...
+}
+```
+
+###  /api/v1/topology/:id/profiling/stop/:host-port (GET)
+
+Request to stop profiler on worker. Returns status and worker id for the request.
+
+|Parameter |Value   |Description  |
+|----------|--------|-------------|
+|id   	   |String (required)| Topology Id  |
```
Comment (revans2) [inline STORM-UI-REST-API.md]: And here

**Comment by revans2** [comment]
The changes look good and I am +1 on merging this in.  I had one minor comment on the documentation, but it is not critical and I am happy to have this merged in, with or without the change.

**Comment by kishorvpatil** [comment]
@revans2  Fixed REST API documentation for consistency. Moved it to right location and added necessary link for Documentation.md

**Comment by revans2** [comment]
Looks good +1

File: storm-core/src/clj/backtype/storm/ui/core.clj (hunk 3 above)
Comment (hustfxj) [inline storm-core/src/clj/backtype/storm/ui/core.clj]: I am concern about Thread concurrency. If user is setting the worker's ProfileAction, supervisor is reading the worker's ProfileAction. Then what happened？

Comment (kishorvpatil) [inline storm-core/src/clj/backtype/storm/ui/core.clj]: As rest of the communication between nimbus and supervisor for assignments - going through ZK answers all Thread concurrency issues. As user ProfileAction is written to ZK by nimbus ( which performs 2 phase commit). The supervisor can read old value or new depending on whether it reads is before or after the commit phase.. The supervisor reads it every 30 seconds.

**Pull request merged**